SignalServiceKit/src/Profiles/OWSProfilesManager.m

```diff
@@ -0,0 +1,144 @@
+NSString *const kOWSProfilesManager_LocalProfileKey = @"kOWSProfilesManager_LocalProfileKey";
+
+// TODO:
+static const NSInteger kProfileKeyLength = 16;
+
+@interface OWSProfilesManager ()
+
+@property (nonatomic, readonly) TSStorageManager *storageManager;
+@property (nonatomic, readonly) OWSMessageSender *messageSender;
+
+@property (nonatomic, readonly, nullable) NSData *localProfileKey;
+
+@end
+
+#pragma mark -
+
+@implementation OWSProfilesManager
+
+@synthesize localProfileKey = _localProfileKey;
+
++ (instancetype)sharedManager
+{
+    static OWSProfilesManager *sharedMyManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedMyManager = [[self alloc] initDefault];
+    });
+    return sharedMyManager;
+}
+
+- (instancetype)initDefault
+{
+    TSStorageManager *storageManager = [TSStorageManager sharedManager];
+    OWSMessageSender *messageSender = [TextSecureKitEnv sharedEnv].messageSender;
+
+    return [self initWithStorageManager:storageManager messageSender:messageSender];
+}
+
+- (instancetype)initWithStorageManager:(TSStorageManager *)storageManager
+                         messageSender:(OWSMessageSender *)messageSender
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(storageManager);
+    OWSAssert(messageSender);
+
+    _storageManager = storageManager;
+    _messageSender = messageSender;
+
+    OWSSingletonAssert();
+
+    // Register this manager with the message sender.
+    // This is a circular dependency.
+    [messageSender setProfilesManager:self];
+
+    // Try to load.
+    _localProfileKey = [self.storageManager objectForKey:kOWSProfilesManager_LocalProfileKey
+                                            inCollection:kOWSProfilesManager_Collection];
+    if (!_localProfileKey) {
+        // Generate
+        _localProfileKey = [OWSProfilesManager generateLocalProfileKey];
+        // Persist
+        [self.storageManager setObject:_localProfileKey
+                                forKey:kOWSProfilesManager_LocalProfileKey
+                          inCollection:kOWSProfilesManager_Collection];
+    }
+    OWSAssert(_localProfileKey.length == kProfileKeyLength);
+
+    return self;
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)observeNotifications
+{
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationDidBecomeActive:)
+                                                 name:UIApplicationDidBecomeActiveNotification
+                                               object:nil];
+}
+
+#pragma mark - Local Profile Key
+
++ (NSData *)generateLocalProfileKey
+{
+    // TODO:
+    OWSFail(@"Profile key generation is not yet implemented.");
+    return [SecurityUtils generateRandomBytes:kProfileKeyLength];
+}
+
+- (nullable NSData *)localProfileKey
```

Comment:
Yup, this is just a first draft of a first draft.  It's already been slightly tweaked in the next PR.